drivers/vulkan/rendering_device_driver_vulkan.cpp

```diff
@@ -5268,6 +5562,374 @@ RDD::PipelineID RenderingDeviceDriverVulkan::render_pipeline_create(
+		uint8_t *data_ptr = buffer_map(accel_info->instances_buffer);
+		ERR_FAIL_NULL_V(data_ptr, AccelerationStructureID());
+		memcpy(data_ptr, accel_info->instances.ptr(), instances_size);
+		buffer_unmap(accel_info->instances_buffer);
+		instances_buffer_address = buffer_get_device_address(accel_info->instances_buffer);
+	}
+
+	accel_info->geometry.sType = VK_STRUCTURE_TYPE_ACCELERATION_STRUCTURE_GEOMETRY_KHR;
+	accel_info->geometry.geometryType = VK_GEOMETRY_TYPE_INSTANCES_KHR;
+	accel_info->geometry.geometry.instances.sType = VK_STRUCTURE_TYPE_ACCELERATION_STRUCTURE_GEOMETRY_INSTANCES_DATA_KHR;
+	accel_info->geometry.geometry.instances.data.deviceAddress = instances_buffer_address;
+
+	accel_info->build_info.sType = VK_STRUCTURE_TYPE_ACCELERATION_STRUCTURE_BUILD_GEOMETRY_INFO_KHR;
+	accel_info->build_info.flags = VK_BUILD_ACCELERATION_STRUCTURE_PREFER_FAST_TRACE_BIT_KHR;
+	accel_info->build_info.geometryCount = 1;
+	accel_info->build_info.pGeometries = &accel_info->geometry;
+	accel_info->build_info.mode = VK_BUILD_ACCELERATION_STRUCTURE_MODE_BUILD_KHR;
+	accel_info->build_info.type = VK_ACCELERATION_STRUCTURE_TYPE_TOP_LEVEL_KHR;
+
+	VkAccelerationStructureBuildSizesInfoKHR size_info = {};
+	size_info.sType = VK_STRUCTURE_TYPE_ACCELERATION_STRUCTURE_BUILD_SIZES_INFO_KHR;
+	vkGetAccelerationStructureBuildSizesKHR(vk_device, VK_ACCELERATION_STRUCTURE_BUILD_TYPE_DEVICE_KHR, &accel_info->build_info, &instance_count, &size_info);
+	accel_info->range_info.primitiveCount = instance_count;
+
+	_acceleration_structure_create(VK_ACCELERATION_STRUCTURE_TYPE_TOP_LEVEL_KHR, size_info, accel_info);
+	return AccelerationStructureID(accel_info);
+#else
+	return AccelerationStructureID();
+#endif
+}
+
+#if !(defined(MACOS_ENABLED) || defined(IOS_ENABLED))
+static VkDeviceAddress _align_up_address(VkDeviceAddress address, VkDeviceAddress alignment) {
+	return (address + (alignment - 1)) & ~(alignment - 1);
+}
+#endif
+
+void RenderingDeviceDriverVulkan::_acceleration_structure_create(VkAccelerationStructureTypeKHR p_type, VkAccelerationStructureBuildSizesInfoKHR p_size_info, AccelerationStructureInfo *r_accel_info) {
+#if !(defined(MACOS_ENABLED) || defined(IOS_ENABLED))
+	RDD::BufferID buffer = buffer_create(p_size_info.accelerationStructureSize, RDD::BUFFER_USAGE_ACCELERATION_STRUCTURE_STORAGE_BIT | RDD::BUFFER_USAGE_STORAGE_BIT | RDD::BUFFER_USAGE_DEVICE_ADDRESS_BIT, RDD::MEMORY_ALLOCATION_TYPE_GPU);
+	r_accel_info->buffer = buffer;
+
+	// Scratch address must be a multiple of minAccelerationStructureScratchOffsetAlignment.
+	uint64_t padded_scratch_size = p_size_info.buildScratchSize + acceleration_structure_capabilities.min_acceleration_structure_scratch_offset_alignment;
+
+	RDD::BufferID scratch_buffer = buffer_create(padded_scratch_size, RDD::BUFFER_USAGE_STORAGE_BIT | RDD::BUFFER_USAGE_DEVICE_ADDRESS_BIT, RDD::MEMORY_ALLOCATION_TYPE_GPU);
```

Comment:
We should find a way to pass this scratch buffer as a parameter to avoid constant allocation when TLAS creation is done.